src/Microsoft.TestPlatform.CrossPlatEngine/TestEngine.cs

```diff
@@ -224,31 +224,61 @@ public TestEngine() : this(TestRuntimeProviderManager.Instance, new ProcessHelpe
             Func<ProxyOperationManager> proxyCreator = () =>
             {
                 var hostManager = this.testHostProviderManager.GetTestHostManagerByRunConfiguration(testSessionCriteria.RunSettings);
-                hostManager?.Initialize(TestSessionMessageLogger.Instance, testSessionCriteria.RunSettings);
+                if (hostManager == null)
+                {
+                    throw new TestPlatformException(
+                        string.Format(
+                            CultureInfo.CurrentCulture,
+                            Resources.Resources.NoTestHostProviderFound));
+                }
 
+                hostManager.Initialize(TestSessionMessageLogger.Instance, testSessionCriteria.RunSettings);
                 if (testSessionCriteria.TestHostLauncher != null)
                 {
                     hostManager.SetCustomLauncher(testSessionCriteria.TestHostLauncher);
                 }
 
-                var requestSender = new TestRequestSender(requestData.ProtocolConfig, hostManager);
+                var requestSender = new TestRequestSender(requestData.ProtocolConfig, hostManager)
+                {
+                    CloseConnectionOnOperationComplete = false
+                };
 
+                // TODO (copoiena): For now we don't support data collection alongside test
+                // sessions.
+                // 
+                // The reason for this is that, in the case of Code Coverage for example, the
```

Comment:
For more details on this I think @jakubch1 would be the best person to ask. From what I remember, we discussed the possibility of using test sessions in conjecture with Code Coverage but that's not really possible. Code Coverage assumes that by the time the binaries are ready the testhosts running the tests haven't been started yet, and so it can instrument the binaries and then set some environment variables (kind of similar to profiling I think). Then, the testhosts will be started and will make use of the environment variables that were set by the CC data collector.

However, with the sessions API you start the testhosts when you issue the `StartTestSession` call and until recently we discussed the possibility of starting the session during build (which is not really possible due to file locking issues for now). Provided we're able to fix this, this behavior would be incompatible with CC.